doc/kubernetes/container-image-management.md

```diff
@@ -0,0 +1,76 @@
+
+So we need to manage the container image to fix these two problems
+
+# Existing designs need to know first
+### 1. local mode and kube mode
+The container has two modes. Kube mode means that the contianer is deployed by k8s, the container image version is defined in daemonset like below. The image need to be downloaded from ACR, so it's with ACR prefix.
+```
+sonick8scue.azurecr.io/docker-sonic-telemetry:v1
+```
+ Local mode means that the container is created and started by systemd service, and systemd service will always use local latest image like below to create container.
+ ```
+ docker-sonic-telemetry:latest
+ ```
+
+### 2. Kube container will not replace the local container when their image versions are the same.
+##### why this design
+when the kube container and local container are in the same version, local container should be more stable, becasue it has run for some time. And when device first join k8s cluster as worker, k8s will try to deploy a same version container with local, if we choose replace, the container will has a down time, actually no upgrade happens, down time is not necessary.
+##### one note
+The kube container will dry-run, no real services start inside the container. But it still downloads the sonick8scue.azurecr.io/docker-sonic-telemetry:v1 image on device.
+# How to handle the two problems
+For the first problem, since the systemd service will alway use image with latest tag, we can tag the newest version image with latest. How to Tag-Latest:
+- When: If the new version of container has been running for more than 10 minutes, it means it's stable, and we will tag it as latest.
+- Our customized k8s controller will complete the upgrade process within 10 minutes, if fallback happens or something wrong, it's impossible that the container run for more than 10 minutes, for this case, we can't tag the new version image to latest
+- After we tag latest, we need to remove the previous stopped local container, otherwise when go back to local, systemd service will not create new local container with new version image and will start the stopped container which running with a previous version image.
+
+For the second problem, we need to remove the old version images in time. How to Clean-Up:
+- When: Tag latest will trigger Clean-Up, once the tag latest successfully finished, we will do Clean-Up
+- Clean-Up will clean up all old version images beside the last latest version image(for fallback usage)
+- For image removal, we have three cases to handle:
+    - Case 1: after device(v1) joined k8s cluster, k8s will deploy v1 container to device, kube v1 container  will dry-run. Due to no local to kube replacement happens, Tag-Latest and Clean-Up will not happen. Then we upgrade container to v2 via k8s, kube v2 container will really run and replace the local one. Then the images on device after Tag-Latest should be like below:
```

Comment:
I'm afraid the term "really run and replace the local one" would be very confusing to people.